yatsm/pipeline/_topology.py

```diff
@@ -111,11 +114,27 @@ def _config_to_deps(config, dsk=None, overwrite=True):
                 logger.debug('Task already computed and not overwrite - not '
                              'adding: {}'.format(task))
 
-        # If this task didn't provide any new data/record, cull it
+        # TODO: maybe we don't want to cull tasks, because they
+        #           * might have important side effects (e.g., IO)
+        #           * overwrite data (e.g., masking provides and
+        #             requires `data: ['nir']`)
         if not task_needed:
-            logger.debug('Culling task {} because everything it provides is '
-                         'already calculated (e.g., from cache)'.format(task))
-            del dsk[task]
+            if not prov:
+                logger.debug('Would cull "{0}" but task does not claim it '
+                             'provides anything so we will assume it has '
+                             'side-effects'.format(task))
+            elif all([p in deps for p in prov]):
```

Comment:
This check is designed to allow preprocessing steps, like applying a mask band or applying min/max valid data range based masks. Using these examples, these tasks should probably "overwrite" data in memory, because we don't usually care about the unmasked data after we've masked. I'm contemplating a few ways of representing this:

* check outputs against requirements to see if it looks like it overwrites, as done here
* have the user specify that a task should overwrite, and check it that way
* have some system where a task declares it should overwrite when defined in code (e.g., similar to the "eager task" concept)

One consequence of having this side-effect accepting  logic opened up is that the user becomes responsible for correctly ordering the tasks in the configuration specification. I imagine most of these tasks that overwrite data do not really care what order the operations are performed in so maybe it doesn't matter as much...